amacc.c

```diff
@@ -509,57 +502,42 @@ void expr(int lev)
         expr(Inc); // dereference has the same precedence as Inc(++)
         if (ty >= PTR) ty -= PTR;
         else fatal("bad dereference");
-        if (ty >= PTR) {
-            *++e = LI;
-        } else if (ty == INT) {
-            *++e = LI;
-        } else if (ty == CHAR) {
-            *++e = LC;
+        if (ty == PTR || ty == INT || ty ==CHAR) {
+            *--n = ty; *--n = Load;
         } else fatal("unexpected type");
         break;
     case And: // "&", take the address operation
         /* when "token" is a variable, it takes the address first and
          * then LI/LC, so `--e` becomes the address of "a".
          */
         next(); expr(Inc);
-        if (*e == LC || *e == LI) --e;
+        if (*n == Load) n += 2; else { fatal("bad address-of");}
         ty += PTR;
         break;
     case '!': // "!x" is equivalent to "x == 0"
         next(); expr(Inc);
-        *++e = PSH; *++e = IMM; *++e = 0; *++e = EQ; ty = INT;
+        if (*n == Num) n[1] = !n[1]; else { *--n = 0; *--n = Num; --n; *n = (int) (n + 3); *--n = Eq; }
+        ty = INT;
         break;
     case '~': // "~x" is equivalent to "x ^ -1"
         next(); expr(Inc);
-        *++e = PSH; *++e = IMM; *++e = -1; *++e = XOR; ty = INT;
+        if (*n == Num) n[1] = ~n[1]; else { *--n = -1; *--n = Num; --n; *n = (int) (n + 3); *--n = Xor; }
+        ty = INT;
         break;
     case Add:
         next(); expr(Inc); ty = INT;
         break;
     case Sub:
-        next(); *++e = IMM;
-        if (tk == Num) { *++e = -ival; next(); } // value, taking negative
-        else { *++e = -1; *++e = PSH; expr(Inc); *++e = MUL; }
+        next(); 
+        expr(Inc);
+        if (*n == Num) n[1] = -n[1]; else { *--n = -1; *--n = Num; --n; *n = (int) (n + 3); *--n = Mul; }
         ty = INT;
         break;
     // processing ++x and --x. x-- and x++ is handled later
     case Inc:
     case Dec:
         t = tk; next(); expr(Inc);
-        switch (*e) {
-        // push address to stack (used by SC/SI below), then take the number
-        case LC: *e = PSH; *++e = LC; break;
-        case LI: *e = PSH; *++e = LI; break;
-        default: fatal("bad lvalue in pre-increment");
-        }
-        // pushing the value
-        // the pointer is added or subtracted, otherwise it is added or subtracted by 1
-        *++e = PSH;
-        *++e = IMM;
-        *++e = ty >= PTR2 ? sizeof(int) :
-                            (ty >= PTR) ? tsize[ty - PTR] : 1;
-        *++e = (t == Inc) ? ADD : SUB;
-        *++e = (ty == CHAR) ? SC : SI;
+        if (*n == Load) *n = t; else { fatal("bad lvalue in pre-increment"); }
```

Comment:
[style] Should be `if (*n == Load) *n = t; else fatal("bad lvalue in pre-increment");`